adminapi/base.py

```diff
@@ -0,0 +1,450 @@
+            if action == 'update':
+                change['old'] = old_value
+                change['new'] = new_value
+            elif action == 'multi':
+                change['remove'] = old_value.difference(new_value)
+                change['add'] = new_value.difference(old_value)
+
+            changes[key] = change
+
+        return changes
+
+    def _confirm_changes(self):
+        self.old_values.clear()
+        if self._deleted:
+            self.object_id = None
+            self._deleted = False
+
+    def _build_commit_object(self):
+        changes = {}
+        if self.is_dirty():
+            changes[self.object_id] = self._serialize_changes()
+        return {
+            'deleted': [self.object_id] if self.is_deleted() else [],
+            'changes': changes,
+        }
+
+    def _save_old_value(self, key):
+        if key not in self.old_values:
+            old_value = self[key]
+            if isinstance(old_value, MultiAttr):
+                self.old_values[key] = old_value.copy()
+            else:
+                self.old_values[key] = old_value
+
+    def __setitem__(self, key, value):
+        if self._deleted:
+            raise DatasetError('Cannot set attributes to deleted server')
+        self.validate(key, value)
+        if isinstance(self[key], MultiAttr):
+            # Multi attributes are guaranteed to exist as MultiAttr, so
+            # we can always get the previous datatype from the existing
+            # value.
+            value = MultiAttr(value, self, key, self[key].datatype)
+        if key not in self.old_values or self.old_values[key] != value:
+            self._save_old_value(key)
+        return super(BaseServerObject, self).__setitem__(key, value)
+
+    def validate(self, key, value):
+        if key not in self:
+            raise KeyError(
+                'Cannot set nonexistent attribute "{}"'.format(key)
+            )
+        if isinstance(self[key], MultiAttr):
+            # We are not extensively validating multi attributes.  Their
+            # values will validated by the MultiAttr class.
+            if not isinstance(value, (tuple, list, set, frozenset)):
+                raise TypeError('Attribute "{}" must be multi'.format(key))
+        elif isinstance(self[key], bool):
+            # Boolean attributes are guaranteed to exist as booleans, so
+            # we can rely on the existing value.
+            if not isinstance(value, bool):
+                raise TypeError('Attribute "{}" must be a boolean'.format(key))
+        elif value is not None:
+            datatype = type(self[key]) if self[key] is not None else None
+            if not datatype and self.old_values.get(key) is not None:
+                datatype = type(self.old_values[key])
+            _validate_value(key, value, datatype)
+
+    def __delitem__(self, key):
+        self[key] = None
+
+    def clear(self):
+        for key in self:
+            del self[key]
```

Comment:
Is it really meant to set every key to `None` or rather to clear also all the keys?

Reply:
I don't think this is used anywhere in serveradmin directly so it's hard to tell, but I don't see when clearing the keys themselves would be usefull. Can you?

Reply:
It doesn't look very useful to me either, given that you would never be able to commit this `Server` back to to Serveradmin.  Some attributes are surely required.  This is client-side code, so we would never know how is relying on it.  We can try removing it, see if anybody would complain.  I am appending another commit for it.